mongo/integration/collection_test.go

```diff
@@ -1762,6 +1760,29 @@ func TestCollection(t *testing.T) {
 				SetUpsert(true),
 			)
 
+			// Seed mock responses. Responses look like:
+			// {ok: 1, n: <matched count>, nModified: <modified count>, upserted: [{index:<index>, _id:<ID>},...]}.
+			var responses []bson.D
+			for i := numModels; i > 0; i -= maxBatchCount {
+				count := maxBatchCount
+				// the first document is upserted
```

Comment:
It took me a little while to realize what this comment meant. But I realize that it is because the updates are in sequence:

- `q: {a: 1}, u: {$set: {a: 2}, upsert: true}`
- `q: {a: 2}, u: {$set: {a: 3}, upsert: true}`
- `q: {a: 3}, u: {$set: {a: 4}, upsert: true}`
- ...

Only the first update in the sequence would do an upsert. The subsequent updates would repeatedly update the first upserted document.

The intent was captured in the comment above `// it's significantly faster to upsert one model and modify the rest than to upsert all of them`.

```suggestion
				// only the first document of each batch is upserted
```

Reply:
It's not actually the first one of each batch, but the first one of the first batch and the last one of the last (second) batch. Changed the comment